src/go/debug-collector/dumper/dumper.go

```diff
@@ -0,0 +1,222 @@
+		"roles",
+		"storageclasses",
+		"persistentvolumeclaims",
+		"persistentvolumes",
+	}
+	if len(resource) > 0 {
+		resources = append(resources, resource)
+	}
+	return Dumper{
+		cmd:       "kubectl",
+		resources: resources,
+		location:  directory,
+		Errors:    make(map[string]string),
+		mode:      int64(0777),
+		namespace: namespace,
+	}
+}
+
+type k8sPods struct {
+	Items []corev1.Pod `json:"items"`
+}
+
+type namespaces struct {
+	Items []corev1.Namespace `json:"items"`
+}
+
+// DumpCluster create dump of a cluster in Dumper.location
+func (d *Dumper) DumpCluster() error {
+	tarFile, err := os.Create(d.location + ".tar.gz")
+	if err != nil {
+		return errors.Wrap(err, "create tar file")
+	}
+	defer tarFile.Close()
+	zr := gzip.NewWriter(tarFile)
+	tw := tar.NewWriter(zr)
+	defer zr.Close()
+	defer tw.Close()
+
+	var nss namespaces
+
+	if len(d.namespace) > 0 {
+		ns := corev1.Namespace{}
+		ns.Name = d.namespace
+		nss.Items = append(nss.Items, ns)
+	} else {
+		output, err := d.runCmd("get", "namespaces", "-o", "json")
+		if err != nil {
+			return errors.Wrap(err, "get namespaces")
+		}
+
+		err = json.Unmarshal(output, &nss)
+		if err != nil {
+			return errors.Wrap(err, "unmarshal namespaces")
+		}
+	}
+
+	for _, ns := range nss.Items {
+		output, err := d.runCmd("get", "pods", "-o", "json", "--namespace", ns.Name)
+		if err != nil {
+			continue // runCmd already stored this error in Dumper.Errors
+		}
+		var pods k8sPods
+		err = json.Unmarshal(output, &pods)
+		if err != nil {
+			d.saveCommandError(err.Error(), "unmarshal pods from namespace", ns.Name)
+			log.Println(errors.Wrap(err, "unmarshal pods"))
+		}
+
+		for _, pod := range pods.Items {
+			location := d.location + "/" + ns.Name + "/" + pod.Name + "/logs.txt"
+			output, err = d.runCmd("logs", pod.Name, "--namespace", ns.Name, "--all-containers")
+			if err != nil {
+				err = createArchive(location, d.mode, []byte(err.Error()), tw)
+				if err != nil {
+					log.Printf("create archive with err: %v", err)
+				}
+				continue
+			}
+			err = createArchive(location, d.mode, output, tw)
+			if err != nil {
+				log.Printf("create archive for pod %s: %v", pod.Name, err)
+			}
+		}
+
+		for _, resource := range d.resources {
+			err = d.getResource(resource, ns.Name, tw)
+			if err != nil {
+				log.Println(errors.Wrapf(err, "get %s resource", resource))
+			}
+		}
+
+		err = d.writeErrorsToFile(tw)
+		if err != nil {
+			log.Println(errors.Wrap(err, "write errors"))
+		}
+	}
+
+	err = d.getResource("nodes", "", tw)
+	if err != nil {
+		log.Println(errors.Wrapf(err, "get nodes"))
+	}
+
+	return nil
+}
+
+// runCmd run command (Dumper.cmd) with given args, return it output and save all errors in Dumper.errors
+func (d *Dumper) runCmd(args ...string) ([]byte, error) {
+	var outb, errb bytes.Buffer
+	cmd := exec.Command(d.cmd, args...)
+	cmd.Stdout = &outb
+	cmd.Stderr = &errb
+	err := cmd.Run()
+	if err != nil {
+		d.saveCommandError(err.Error()+" "+outb.String(), args...)
```

Comment:
why here is stdout but no stderr?
shouldn't we just return everything in case of an error? Like:
```go
	if err != nil || errb.Len() > 0 {
		return nil, errors.Errorf("error: %v, stderr: %s, stdout: %s", err, errb, outb)
	}

	return outb.Bytes(), nil
```

Reply:
https://github.com/percona/percona-toolkit/pull/452/commits/118110a671ae050d9ba09342b631a0fecf9a964e